neutron_plugin/router.py

```diff
@@ -287,3 +312,126 @@ def _get_connected_ext_net_id(neutron_client):
                 ext_net_ids))
 
     return ext_net_ids[0] if ext_net_ids else None
+
+
+def _update_router_routes(neutron_client, args, **kwargs):
+
+    from copy import deepcopy
+
+    def dict_merge(a, b):
+        if isinstance(a, list) and isinstance(b, list):
+            a.extend(b)
+            return a
+        if not isinstance(b, dict):
+            return b
+        result = deepcopy(a)
+        for k, v in b.iteritems():
+            if k in result:
+                ctx.logger.info('Match {0}'.format(k))
+                result[k] = dict_merge(result[k], v)
+                ctx.logger.info('Match Routes {0}'.format(k))
+        return result
+
+    # Find out if the update script is being called
+    # from a relationship or a node operation.
+    router = _get_router_from_relationship(neutron_client)
+
+    router_id = router[ROUTER_OPENSTACK_TYPE].pop('id')
+    new_router = {ROUTER_OPENSTACK_TYPE: {}}
+    for key, value in args.items():
+        new_router['router'][key] = value
+
+    for ro_attribute in ['status', 'tenant_id']:
+        try:
+            del router[ROUTER_OPENSTACK_TYPE][ro_attribute]
+        except KeyError:
+            pass
+
+    new_router = dict_merge(new_router, router)
+    return neutron_client.update_router(router_id, new_router)
+
+
+def _get_router_from_relationship(neutron_client):
+    # Find out if the update script is being called
+    # from a relationship or a node operation.
+
+    # Only get the "router_rel" if it is not a relationship instance
+    if ctx.type != RELATIONSHIP_INSTANCE:
+        router_rel = get_relationships_by_relationship_type(
+            ctx, ROUTES_OPENSTACK_RELATIONSHIP)
+
+    if ctx.type == RELATIONSHIP_INSTANCE:
+        if ROUTER_OPENSTACK_TYPE in get_openstack_type(ctx.source):
+            subject = ctx.source
+        elif ROUTER_OPENSTACK_TYPE in get_openstack_type(ctx.target):
+            subject = ctx.target
+        else:
+            raise NonRecoverableError(
+                'Neither target nor source is {0}'.format(
+                    ROUTER_OPENSTACK_TYPE))
+
+    elif router_rel and ROUTER_OPENSTACK_TYPE\
+            in get_openstack_type(router_rel[0].target):
+
+        subject = router_rel[0].target
+
```

Comment:
The code is kind of confusing like this. I would prefer to see all non `RELATIONSHIP_INSTANCE` logic in one conditional like:
```
if ctx.type != RELATIONSHIP_INSTANCE:
   router_rel = ...
   if not ROUTER_OPENSTACK_TYPE  in get_openstack_type(router_rel[0].target):
       subject = ctx
   subject = router_rel[0].target
```
Otherwise, it obscures that this is really a sub-condition of a previous condition.

Reply:
I think this make sense I will update this